lib/config.cpp

```diff
@@ -705,23 +705,31 @@ json Config::resolve_requirement(const std::string& module_id, const std::string
     return module_config["connections"][requirement_id];
 }
 
-std::list<Requirement> Config::get_requirements(const std::string& module_id) const {
+std::map<std::string, std::vector<Fulfillment>> Config::get_fulfillments(const std::string& module_id) {
```

Comment:
As mentioned above, I would not add `mapping` to the `Requirement` struct, and hence this function and also the functions using this one, would just stay the same.  Better create the above mentioned struct (which contains the mapping `Requirement/Module -> mapping`) either on the fly or precompute it and then just return that.